pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,781 @@
+                new_domains.append(d.add_constraint(isl.Constraint.eq_from_names(d.space, {"n": 1, "start": -1})))
+            else:
+                new_domains.append(d)
+        domains = new_domains
+        if builder.extruded:
+            new_domains = []
+            for d in domains:
+                if d.get_dim_name(isl.dim_type.set, 0) == builder.layer_index.name:
+                    # layer = t1 - 1
+                    t1 = parameters.layer_end
+                    new_domains.append(d.add_constraint(isl.Constraint.eq_from_names(d.space, {"layer": 1, t1: -1, 1: 1})))
+                else:
+                    new_domains.append(d)
+        domains = new_domains
+
+    assumptions, = reduce(operator.and_,
+                          parameters.assumptions.values()).params().get_basic_sets()
+    options = loopy.Options(check_dep_resolution=True, ignore_boostable_into=True)
+
+    # sometimes masks are not used, but we still need to create the function arguments
+    for i, arg in enumerate(parameters.wrapper_arguments):
+        if parameters.kernel_data[i] is None:
+            arg = loopy.GlobalArg(arg.name, dtype=arg.dtype, shape=arg.shape)
+            parameters.kernel_data[i] = arg
+
+    if wrapper_name is None:
+        wrapper_name = "wrap_%s" % builder.kernel.name
+
+    wrapper = loopy.make_kernel(domains,
+                                statements,
+                                kernel_data=parameters.kernel_data,
+                                target=loopy.CTarget(),
+                                temporary_variables=parameters.temporaries,
+                                symbol_manglers=[symbol_mangler],
+                                options=options,
+                                assumptions=assumptions,
+                                lang_version=(2018, 2),
+                                name=wrapper_name)
+
+    # additional assumptions
+    if builder.single_cell:
+        wrapper = loopy.assume(wrapper, "start < end")
+    else:
+        wrapper = loopy.assume(wrapper, "start <= end")
+    wrapper = loopy.assume(wrapper, "start >= 0")
+    if builder.extruded:
+        wrapper = loopy.assume(wrapper, "{0} <= {1}".format(parameters.layer_start, parameters.layer_end))
+
+    # prioritize loops
+    for indices in context.index_ordering:
+        wrapper = loopy.prioritize_loops(wrapper, indices)
+
+    # register kernel
+    kernel = builder.kernel
+    headers = set(kernel._headers)
+    headers = headers | set(["#include <math.h>"])
+    preamble = "\n".join(sorted(headers))
+
+    from coffee.base import Node
+
+    if isinstance(kernel._code, loopy.LoopKernel):
+        knl = kernel._code
+        wrapper = loopy.register_callable_kernel(wrapper, knl)
+        from loopy.transform.callable import _match_caller_callee_argument_dimension_
+        wrapper = _match_caller_callee_argument_dimension_(wrapper, knl.name)
+        wrapper = loopy.inline_callable_kernel(wrapper, knl.name)
```

Comment:
@inducer/@kaushikcfd, do we always have to inline the kernel if it is a loopy kernel? Also, this matching of the argument dimensions, what do we need to change in the kernel generation that it is unneeded?

Reply:
1. No, if the inline transformation is not called then separate functions are generated and the corresponding arguments are passed.
2. As far as I remember, there is a mismatch in the dimensions of the "coordinates" argument between the caller and callee(for ex. in tsfc it may have shape`(6, )` and in pyop2 `(3, 2)`) . And that's why we need `_match_caller_callee_argument_dimension_` which changes the shape of the arguments of the callee kernel assuming C-ordering of the data layout.

Reply:
OK, so if I remove the `inline_callable_kernel` line, then everything should continue to work?

Reply:
Yes.

Reply:
I get:
```python
    def generate_code_v2(program):
        """
        Returns an instance of :class:`CodeGenerationResult`.
    
        :param program: An instance of :class:`loopy.Program`.
        """
        from loopy.kernel import LoopKernel
        from loopy.program import make_program
    
        if isinstance(program, LoopKernel):
            program = make_program(program)
    
        from loopy.kernel import KernelState
        if program.root_kernel.state == KernelState.INITIAL:
            from loopy.preprocess import preprocess_program
            program = preprocess_program(program)
    
        from loopy.type_inference import infer_unknown_types
        program = infer_unknown_types(program, expect_completion=True)
    
        codegen_results = {}
    
        for func_id, in_knl_callable in program.callables_table.items():
            if isinstance(in_knl_callable, CallableKernel):
                codegen_results[func_id] = (
                        generate_code_for_a_single_kernel(in_knl_callable.subkernel,
                            program.callables_table, program.target))
                if not in_knl_callable.subkernel.is_called_from_host:
                    assert codegen_results[func_id].host_program is None
    
        device_preambles = []
        for cgr in codegen_results.values():
            device_preambles.extend(cgr.device_preambles)
    
        # collecting the function declarations of callee kernels
        for in_knl_callable in program.callables_table.values():
            for preamble in in_knl_callable.generate_preambles(program.target):
                device_preambles.append(preamble)
    
        collective_device_program = codegen_results[program.name].device_programs[0]
        callee_fdecls = []
    
        for func_id, callee_cgr in codegen_results.items():
            if func_id != program.name:
                assert len(callee_cgr.device_programs) == 1
                callee_prog_ast = callee_cgr.device_programs[0].ast
                collective_device_program = collective_device_program.copy(
                        ast=Collection([callee_prog_ast, collective_device_program.ast]))
>               callee_fdecls.append(callee_prog_ast.fdecl)
E               AttributeError: 'Collection' object has no attribute 'fdecl'

../loopy/loopy/codegen/__init__.py:623: AttributeError```

It looks like this code assumes that the kernel has a single thing which is a functionbody, but that doesn't handle the case we have where we have a `Collection`.

If I try the obvious patch:
```
diff --git a/loopy/codegen/__init__.py b/loopy/codegen/__init__.py
index 8f3e15f2..1e5ff8b1 100644
--- a/loopy/codegen/__init__.py
+++ b/loopy/codegen/__init__.py
@@ -620,7 +620,14 @@ def generate_code_v2(program):
             callee_prog_ast = callee_cgr.device_programs[0].ast
             collective_device_program = collective_device_program.copy(
                     ast=Collection([callee_prog_ast, collective_device_program.ast]))
-            callee_fdecls.append(callee_prog_ast.fdecl)
+            if isinstance(callee_prog_ast, Collection):
+                for entry in callee_prog_ast.contents:
+                    try:
+                        callee_fdecls.append(entry.fdecl)
+                    except AttributeError:
+                        pass
+            else:
+                callee_fdecls.append(entry.fdecl)
 
     # collecting the function declarations of callee kernels
     for callee_fdecl in callee_fdecls:
```

then I get a further error (somehow some arguments with shape (1, ) have been rewritten to shape ().

I have an inner kernel:
```
---------------------------------------------------------------------------
KERNEL: expression
---------------------------------------------------------------------------
ARGUMENTS:
fn_0: type: np:dtype('float64'), shape: (1), dim_tags: (N0:stride:1) aspace: global
fn_1: type: np:dtype('float64'), shape: (1), dim_tags: (N0:stride:1) aspace: global
fn_2: type: np:dtype('float64'), shape: (1), dim_tags: (N0:stride:1) aspace: global
---------------------------------------------------------------------------
DOMAINS:
{ [d] : d = 0 }
---------------------------------------------------------------------------
INAME IMPLEMENTATION TAGS:
d: None
---------------------------------------------------------------------------
INSTRUCTIONS:
for d
  fn_0[d] = fn_1[d] + (-1)*fn_2[d]  {id=insn}
end d
------
```

And an outer kernel:
```
---------------------------------------------------------------------------
KERNEL: wrap_expression
---------------------------------------------------------------------------
ARGUMENTS:
dat0: type: np:dtype('float64'), shape: (None), dim_tags: (N0:stride:1) aspace: global
dat1: type: np:dtype('float64'), shape: (None), dim_tags: (N0:stride:1) aspace: global
dat2: type: np:dtype('float64'), shape: (None), dim_tags: (N0:stride:1) aspace: global
end: ValueArg, type: np:dtype('int32')
start: ValueArg, type: np:dtype('int32')
---------------------------------------------------------------------------
DOMAINS:
[end, start] -> { [n] : start <= n < end }
---------------------------------------------------------------------------
INAME IMPLEMENTATION TAGS:
n: None
---------------------------------------------------------------------------
INSTRUCTIONS:
for n
  SubArrayRef((), (dat0[n])) = expression(SubArrayRef((), (dat1[n])), SubArrayRef((), (dat2[n])))  {id=statement0}
end n
```
```
But I get an error:

            if len(subscript) != len(shape):
                raise LoopyError("subscript to '%s' in '%s' has the wrong "
                        "number of indices (got: %d, expected: %d)" % (
                            expr.aggregate.name, expr,
>                           len(subscript), len(shape)))
E               loopy.diagnostic.LoopyError: subscript to 'fn_0' in 'fn_0[d]' has the wrong number of indices (got: 1, expected: 0)
```

And indeed, something has rewritten the inner kernel into:
```
---------------------------------------------------------------------------
KERNEL: expression
---------------------------------------------------------------------------
ARGUMENTS:
fn_0: type: np:dtype('float64'), shape: () aspace: global
fn_1: type: np:dtype('float64'), shape: () aspace: global
fn_2: type: np:dtype('float64'), shape: () aspace: global
---------------------------------------------------------------------------
DOMAINS:
{ [d] : d = 0 }
---------------------------------------------------------------------------
INAME IMPLEMENTATION TAGS:
d: None
---------------------------------------------------------------------------
INSTRUCTIONS:
for d
  fn_0[d] = fn_1[d] + (-1)*fn_2[d]  {id=insn}
end d```

Reply:
Let me fix it.

Reply:
Thanks!

Reply:
- The problem was the support of empty subarray refs. That has been fixed and pushed to `firedrakeproject/loopy`.
- Corresponding test: https://gitlab.tiker.net/inducer/loopy/blob/firedrake/test/test_callables.py#L449 (tests both with and without inline)

Reply:
OK, this stuff is now fixed, thanks!